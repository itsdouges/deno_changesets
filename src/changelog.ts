import { join } from 'https://deno.land/std@0.170.0/path/mod.ts';
import {
  Change,
  Changelog,
  parser,
  Release,
} from 'https://deno.land/x/changelog@v2.0.0/mod.ts';
import { format } from './deno.ts';
import { Changeset, ChangeType } from './types.ts';

function sortByModule(changesets: Changeset[]) {
  const changeMap: Record<
    string,
    {
      name: string;
      path: string;
      changes: { type: ChangeType; description: string }[];
    }
  > = {};

  for (const { description, modules } of changesets) {
    for (const module of modules) {
      if (!changeMap[module.name]) {
        changeMap[module.name] = {
          name: module.name,
          changes: [],
          path: module.path,
        };
      }

      changeMap[module.name].changes.push({
        type: module.changeType,
        description,
      });
    }
  }

  return Object.values(changeMap);
}

export async function upsert(
  changesets: Changeset[],
  nextVersion: string,
  { __dryRun = false, __date = new Date() } = {},
) {
  const changelogs: {
    md: string;
    path: string;
    action: 'update' | 'create';
  }[] = [];

  const moduleChanges = sortByModule(changesets);

  for (const module of moduleChanges) {
    const path = join(module.path, 'CHANGELOG.md');
    const absolutePath = join(Deno.cwd(), path);
    let changelog: Changelog;
    let action: 'update' | 'create';

    try {
      const md = await Deno.readTextFile(absolutePath);
      changelog = parser(md);
      action = 'update';
    } catch (_) {
      changelog = new Changelog(module.name);
      action = 'create';
    }

    const release = new Release(
      nextVersion,
      __date,
    );

    for (const change of module.changes) {
      release.addChange(change.type, new Change(change.description));
    }

    changelog.addRelease(
      release,
    );

    const md = await format(changelog.toString(), 'md');

    changelogs.push({
      md,
      action,
      path,
    });

    if (__dryRun) {
      // Skip persisting to the filesystem.
      continue;
    }

    await Deno.writeTextFile(absolutePath, md);
  }

  return changelogs;
}
