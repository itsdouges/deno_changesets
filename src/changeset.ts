import { join } from 'https://deno.land/std@0.170.0/path/mod.ts';
import HumanHasher from 'npm:humanhash@1.0.4';
import fm from 'npm:front-matter@4.0.2';
import * as modules from './modules.ts';
import { Changeset, ChangeType, changeTypes } from './types.ts';
import * as git from './git.ts';

const frontmatter = fm as unknown as typeof fm.default;

export function _buildChangeset(
  modules: { name: string; changeType: ChangeType }[],
  description: string,
): string {
  return `---
${modules.map((mod) => `'${mod.name}': ${mod.changeType}`).join('\n')}
---

${description}
`;
}

export async function changeset(path: string) {
  const humanhash = new HumanHasher();
  const repository = await modules.list(path);
  const outputFolder = join(path, '.changeset');

  function moduleInvariant(name: string) {
    if (!repository.modules.find((mod) => mod.name === name)) {
      throw new Error('invariant: module not found');
    }
  }

  function changeTypeInvariant(changeType: ChangeType) {
    if (!changeTypes.includes(changeType)) {
      throw new Error(`invariant: changeType must be one of ${changeTypes}`);
    }
  }

  return {
    create: async (
      modules: { name: string; changeType: ChangeType }[],
      description: string,
      { commit = false } = {},
    ) => {
      if (commit) {
        const dirty = await git.isDirty();
        if (dirty) {
          throw new Error('invariant: dirty git');
        }
      }

      modules.forEach(({ name }) => moduleInvariant(name));

      const content = _buildChangeset(modules, description);
      const filename = humanhash.humanize(content, 3, '_') + '.md';

      await Deno.mkdir(outputFolder, { recursive: true });

      await Deno.writeTextFile(
        join(outputFolder, filename),
        content,
        { createNew: true },
      );

      if (commit) {
        const moduleNames = modules.map((mod) => mod.name).join(', ');

        await git.add();
        await git.commit(`changeset: [${moduleNames}] â€” ${description}`);
      }
    },
    readAll: async (): Promise<Changeset[]> => {
      const changesets: Changeset[] = [];

      for await (const dirEntry of Deno.readDir(outputFolder)) {
        if (!dirEntry.isFile) {
          throw new Error('invariant');
        }

        const contents = await Deno.readTextFile(
          join(outputFolder, dirEntry.name),
        );
        const parsed = frontmatter<Record<string, ChangeType>>(contents);
        const moduleMap = repository.modules.reduce<
          Record<string, modules.Module>
        >((acc, obj) => Object.assign(acc, { [obj.name]: obj }), {});

        changesets.push({
          description: parsed.body.trim(),
          modules: Object.entries(parsed.attributes).map(
            ([name, changeType]) => {
              moduleInvariant(name);
              changeTypeInvariant(changeType);

              return {
                name,
                changeType,
                path: moduleMap[name].path,
              };
            },
          ),
        });
      }

      const changedModules = changesets.flatMap((cs) => cs.modules).map((cs) =>
        cs.name
      );

      const changedModulesDependants = await modules.dependencies(
        path,
        changedModules,
      );

      changedModulesDependants.forEach((mod) => {
        mod.dependencies.forEach((dep) => {
          changesets.push({
            description: `Upgraded \`${dep}\` to @{nextVersion}`,
            modules: [{
              changeType: 'fixed',
              name: mod.moduleName,
              path: mod.path,
            }],
          });
        });
      });

      return changesets;
    },
    deleteAll: async () => {
      await Deno.remove(join(path, '.changeset'), { recursive: true });
    },
  };
}
