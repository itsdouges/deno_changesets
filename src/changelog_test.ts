import { join } from 'https://deno.land/std@0.170.0/path/mod.ts';
import { assertSnapshot } from 'https://deno.land/std@0.170.0/testing/snapshot.ts';
import { upsert } from './changelog.ts';
import { changeset } from './changeset.ts';

Deno.test(async function shouldCreateChangelog(t) {
  const path = join(Deno.cwd(), 'src/__mocks__/changeset_patch');
  const manager = await changeset(path);
  const changesets = await manager.readAll();

  const actual = await upsert(changesets, '0.0.2', {
    __dryRun: true,
    __date: new Date('2000-01-01'),
  });

  await assertSnapshot(t, actual);
});

Deno.test(async function shouldUpdateChangelog(t) {
  const path = join(Deno.cwd(), 'src/__mocks__/changeset_major');
  const manager = await changeset(path);
  const changesets = await manager.readAll();

  const actual = await upsert(changesets, '1.0.0', {
    __dryRun: true,
    __date: new Date('2000-10-01'),
  });

  await assertSnapshot(t, actual);
});
