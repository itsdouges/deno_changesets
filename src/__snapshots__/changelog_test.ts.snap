export const snapshot = {};

snapshot[`shouldCreateChangelog 1`] = `
[
  {
    action: "create",
    md: "# deno_changesets

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this
project adheres to [Semantic Versioning](http://semver.org/).

## 0.0.2 - 2000-01-01

### Deprecated

- Deprecated an API, sorry!
",
    path: "/src/__mocks__/changeset_patch/CHANGELOG.md",
  },
]
`;

snapshot[`shouldUpdateChangelog 1`] = `
[
  {
    action: "update",
    md: "# deno_changesets

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this
project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0 - 2000-10-01

### Added

- Added a new API, yay!

### Removed

- Removed an API, sorry!

### Fixed

- Fixed a bug, woo!

## 0.0.2 - 2000-01-01

### Deprecated

- Deprecated the deprecation, sorry!
",
    path: "/src/__mocks__/changeset_major/CHANGELOG.md",
  },
]
`;
